File: js/scripts/parse-docs.mjs
/**
 * Parse all .mdx files from the docs repo through the Rust pipeline via NAPI.
 *
 * Usage: node scripts/parse-docs.mjs
 */

import { readFileSync, readdirSync, statSync } from 'node:fs';
import { join, relative } from 'node:path';
import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);
const native = require('../tryckeri_napi.linux-x64-gnu.node');

const DOCS_DIR = '/home/erika/Projects/docs';

function collectMdxFiles(dir, out = []) {
  for (const entry of readdirSync(dir)) {
    const full = join(dir, entry);
    try {
      const stat = statSync(full);
      if (stat.isDirectory()) {
        collectMdxFiles(full, out);
      } else if (entry.endsWith('.mdx')) {
        out.push(full);
      }
    } catch {
      // skip
    }
  }
  return out;
}

const files = collectMdxFiles(DOCS_DIR).sort();
console.log(`Found ${files.length} .mdx files\n`);

let success = 0;
let failures = [];
let totalBytes = 0;

const start = performance.now();

for (const path of files) {
  const source = readFileSync(path, 'utf8');
  totalBytes += Buffer.byteLength(source);

  try {
    // Full pipeline: parse MDX → HAST → HTML
    const html = native.parseMdxToHtml(source);
    if (typeof html !== 'string') throw new Error('not a string');
    success++;
  } catch (e) {
    failures.push({ path: relative(DOCS_DIR, path), error: e.message });
  }
}

const elapsed = performance.now() - start;
const mb = totalBytes / (1024 * 1024);

console.log('Results:');
console.log(`  Success: ${success}/${files.length}`);
console.log(`  Failed:  ${failures.length}`);
console.log(`  Total:   ${mb.toFixed(2)} MB in ${elapsed.toFixed(0)}ms`);
console.log(`  Speed:   ${(mb / (elapsed / 1000)).toFixed(1)} MB/s`);

if (failures.length > 0) {
  console.log('\nFailures:');
  for (const f of failures.slice(0, 20)) {
    console.log(`  ${f.path} — ${f.error}`);
  }
  if (failures.length > 20) {
    console.log(`  ... and ${failures.length - 20} more`);
  }
}

File: crates/parser/examples/parse_docs.rs
//! Parse all .mdx files from the docs repo and report results.

use std::path::PathBuf;
use std::time::Instant;

fn main() {
    let docs_dir = PathBuf::from("/home/erika/Projects/docs");
    let opts = parser::ParseOptions::mdx();

    let mut files = Vec::new();
    collect_mdx_files(&docs_dir, &mut files);
    files.sort();

    println!("Found {} .mdx files", files.len());

    let mut success = 0;
    let mut failures: Vec<(PathBuf, String)> = Vec::new();
    let mut total_bytes = 0u64;

    let start = Instant::now();

    for path in &files {
        let source = match std::fs::read_to_string(path) {
            Ok(s) => s,
            Err(e) => {
                failures.push((path.clone(), format!("read error: {e}")));
                continue;
            }
        };
        total_bytes += source.len() as u64;

        // Catch panics.
        let result = std::panic::catch_unwind(|| {
            let arena = parser::parse(&source, &opts);
            // Also try the full pipeline to HTML.
            let _html = tryckeri_hast::arena_to_html(&arena);
            arena.len()
        });

        match result {
            Ok(node_count) => {
                success += 1;
                let _ = node_count;
            }
            Err(e) => {
                let msg = if let Some(s) = e.downcast_ref::<String>() {
                    s.clone()
                } else if let Some(s) = e.downcast_ref::<&str>() {
                    s.to_string()
                } else {
                    "unknown panic".to_string()
                };
                failures.push((path.clone(), msg));
            }
        }
    }

    let elapsed = start.elapsed();

    println!("\nResults:");
    println!("  Success: {}/{}", success, files.len());
    println!("  Failed:  {}", failures.len());
    println!("  Total:   {:.2} MB in {:.2?}", total_bytes as f64 / 1_048_576.0, elapsed);
    println!("  Speed:   {:.2} MB/s", total_bytes as f64 / 1_048_576.0 / elapsed.as_secs_f64());

    if !failures.is_empty() {
        println!("\nFailures:");
        for (path, msg) in &failures {
            let rel = path.strip_prefix(&docs_dir).unwrap_or(path);
            println!("  {} — {}", rel.display(), msg);
        }
    }
}

fn collect_mdx_files(dir: &PathBuf, out: &mut Vec<PathBuf>) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            collect_mdx_files(&path, out);
        } else if path.extension().map_or(false, |e| e == "mdx") {
            out.push(path);
        }
    }
}
